compiler/daml-lf-proto/src/DA/Daml/LF/Proto3/EncodeV1.hs

```diff
@@ -31,14 +32,29 @@ import qualified Proto3.Suite as P (Enumerated (..))
 -- otherwise always be wrapped in `Just` at their call sites.
 type Just a = Maybe a
 
-newtype Encode a = Encode{runEncode :: EncodeEnv -> a}
-    deriving (Functor, Applicative, Monad, MonadReader EncodeEnv)
+type Encode a = State EncodeEnv a
+
+data InternMode = InternNothing | InternPackageIdsOnly
 
 data EncodeEnv = EncodeEnv
     { _version :: Version
-    , internedPackageIds :: S.Set PackageId
+    , internMode :: InternMode
+    , internedStrings :: HMS.HashMap T.Text Word64
     }
 
+initEncodeEnv :: Version -> InternMode -> EncodeEnv
+initEncodeEnv version mode = EncodeEnv version mode HMS.empty
+
+allocString :: T.Text -> Encode Word64
+allocString t = do
+    env@EncodeEnv{internedStrings} <- get
+    case t `HMS.lookup` internedStrings of
+        Just n -> pure n
+        Nothing -> do
+            let n = fromIntegral (HMS.size internedStrings)
+            put $ env{internedStrings = HMS.insert t n internedStrings}
```

Comment:
```suggestion
            put $! env{internedStrings = HMS.insert t n internedStrings}
```
Not a big deal right now since you’ll force it on the next lookup but I prefer to not rely on that kind of stuff since it can easily change during a refactoring.